### ISSUE TITLE
Missing normal must be calculated as flat normals

### ISSUE THREAD

**Issue by asus4** [atteneder/glTFast#340]
**Describe the bug**

First of all, thank you for continuing to develop such a great library.

I found some rendering differences between glTFast and web-viewers. It seems to happen when normal is not included in glTF/glb file.

![compare](https://user-images.githubusercontent.com/357497/155541776-2863f2a0-e80c-46ed-b864-704332a8d5ac.jpg)

This is my export setting in Blender. It seems to only happen if I exported glTF without normal. And other cases worked well.

![export-setting](https://user-images.githubusercontent.com/357497/155541647-32e410c1-02af-4c86-a438-2f33dff8a905.png)

**Files**

[normal-test.zip](https://github.com/atteneder/glTFast/files/8133267/normal-test.zip)

Following files are included:
- test-no-normal.glb: problemic file
- test-with-normal-without-tangent.glb: this was ok 
- normal-test.blend: original blender file

**To Reproduce**
Import the attached file to Unity.

**Expected behavior**

Tangent should be not smoothed if there are no normals included in glTF.

**Desktop (please complete the following information):**
 - glTFast version: 4.6.0
 - Unity Editor version: 2020.3.29f1
 - Render Pipeline and version: URP 10.8.1
 - Platform: macOS Editor

**Comment by atteneder** [comment]
@asus4 Thanks for reporting. I can confirm that this a bug. The [glTF spec](https://www.khronos.org/registry/glTF/specs/2.0/glTF-2.0.html#meshes-overview) says:

> When normals are not specified, client implementations MUST calculate flat normals and the provided tangents (if present) MUST be ignored.

glTFast uses [`Mesh.RecalculateNormals`](https://docs.unity3d.com/2022.1/Documentation/ScriptReference/Mesh.RecalculateNormals.html) to calculate missing normals. It considers all all connected vertices of a vertex, which does something different in case a vertex is assigned to more than one triangle (of different face-normal).

I selected split edges on your test icosphere and as you can see, some vertices share more than one triangle:

<img width="447" alt="image" src="https://user-images.githubusercontent.com/2593014/155588801-5de3459c-8422-4ea6-a2a5-2dc7f30f53fa.png">

The solution would be to:

- Iterate all triangles and vertices
- Duplicate/split verts that have more than one triangle assigned
- Calculate flat normals
- Discard existing tangents and recalculate them

Or alternatively (and probably faster)

- Allocate a new vertex buffer with 3 verts per face (likely more than input verts)
- iterate triangles
- copy per-triangle input vertices to new vertex buffer

As you can see, this is a bit more complicated and that's why I dodged the topic so far.

**Comment by atteneder** [comment]
On another note, again from the spec:

> When normals are not specified, client implementations MUST calculate flat normals and the provided tangents (if present) MUST be ignored.

Given this information it makes no sense that Blender's export settings allow you to not export normals but DO export tangents.

**Comment by asus4** [comment]
@atteneder Thanks for looking into it. 

I had this issue when supporting the C4D default exporter. I guess it has similar options.
https://storage.opensea.io/files/91525ba8a789da1b79155ff63c3c3c5a.gltf
https://storage.opensea.io/files/336992a385208bb9d2e8675f05f47140.gltf

In Three.js GLTF loader, it seems to override normals in the shader (with flat-shading option) instead of modifying the mesh itself. Would it be simpler?

**Comment by atteneder** [comment]
> In Three.js GLTF loader, it seems to override normals in the shader (with flat-shading option) instead of modifying the mesh itself. Would it be simpler?

Yes, it would be. I did not consider this option though, because Unity compiles shaders at build-time. Offering this option would double the number of shader variants, which is already very high.
